File: ios/Storehop/Sync/PullState.swift
import Foundation

/// Lifecycle of a Firestore pull for a given uid. Drives the SyncEngine's
/// push gate (push only runs when `SUCCEEDED`) and the Settings cloud-sync
/// banner (visible when `FAILED`).
///
/// Mirrors Android `PullState`. Same names so Firestore documents under
/// `users/{uid}/_pullState` are wire-compatible if we ever sync the pull
/// state itself (we don't today, but the option is open).
enum PullState: String, Codable, Sendable, CaseIterable {
    /// Initial state for a uid we haven't seen sync for yet. Either it's
    /// never been observed, or pull was attempted but the device was
    /// offline.
    case needed = "NEEDED"
    /// Pull is currently running. Push is paused until this resolves.
    case inProgress = "IN_PROGRESS"
    /// Pull completed successfully. Push is unblocked.
    case succeeded = "SUCCEEDED"
    /// Pull failed. Push stays paused; Settings shows a Retry banner.
    case failed = "FAILED"
}

File: app/src/main/java/com/storehop/app/sync/PullState.kt
package com.storehop.app.sync

/**
 * Per-uid sync state. Drives whether [SyncEngine] is allowed to push for a
 * given uid (only when [SUCCEEDED]).
 *
 * Persisted by [PullStateRepository] in DataStore so process death between
 * sign-in and pull completion doesn't strand the app — on next launch we
 * see [NEEDED] and run the pull again.
 */
enum class PullState {
    /**
     * Initial state for a uid we've never seen before, or after process death
     * mid-pull. The next reachable code path that observes this state should
     * trigger a pull.
     */
    NEEDED,

    /** Pull is currently running. Push side is paused. */
    IN_PROGRESS,

    /**
     * Pull completed successfully or wasn't needed (cloud was empty and the
     * orphan-claim path ran instead). Push side is unblocked.
     */
    SUCCEEDED,

    /**
     * Pull failed (network, auth, deserialization). Push side stays paused so
     * local seeded data can't leak to the cloud. The user sees a Retry banner
     * in Settings.
     */
    FAILED,
}
